torch/fx/experimental/proxy_tensor.py

```diff
@@ -0,0 +1,207 @@
+        # Commenting this out for now since it causes some spurious failures (such as error checking)
+        if func == aten._local_scalar_dense:
+            raise RuntimeError("It appears that you're trying to get value out of a tracing tensor - erroring out! "
+                               "It's likely that this is caused by data-dependent control flow or similar.")
+
+        def unwrap_proxy(e):
+            return e.proxy if isinstance(e, ProxyTensor) else e
+
+        proxy_args = pytree.tree_map(unwrap_proxy, args)
+        proxy_kwargs = pytree.tree_map(unwrap_proxy, kwargs)
+
+        proxy_out = func(*proxy_args, **proxy_kwargs)
+
+        # Kind of a hacky way to test if an op is in-place or not
+        if func.__name__[-1] == "_" and func.__name__[0] != "_":
+            args[0].proxy = proxy_out
+            proxy_out.node.meta['tensor_meta'] = _extract_tensor_metadata(args[0])
+
+        with no_dispatch():
+            real_out = func_overload(*args, **kwargs)
+
+        def wrap_with_proxy(e, proxy):
+            # Some ops (like native_batch_norm_backward) return undefined tensors that get
+            # converted into None in python.
+            # As the function signature expects tensors, if we directly return these None
+            # tensors back to C++, we'll error.
+            if e is None:
+                e = torch.empty(())
+            if type(e) == torch.Tensor:
+                return ProxyTensor(e, proxy)
+            else:
+                return e
+        if isinstance(real_out, tuple):
+            return tuple(wrap_with_proxy(e, proxy_out[idx]) for idx, e in enumerate(real_out))
+        elif isinstance(real_out, list):
+            return list([wrap_with_proxy(e, proxy_out[idx]) for idx, e in enumerate(real_out)])
+        elif isinstance(real_out, torch.Tensor):
+            return wrap_with_proxy(real_out, proxy_out)
+        else:
+            return real_out
+
+
+class PythonKeyTracer(Tracer):
+    def __init__(self):
+        super().__init__()
+
+    def call_module(
+        self, m: torch.nn.Module, forward: Callable[..., Any], args: Tuple[Any, ...], kwargs: Dict[str, Any]
+    ) -> Any:
+        return forward(*args, **kwargs)
+
+    def _module_getattr(self, attr, attr_val, parameter_proxy_cache):
+        if isinstance(attr_val, torch.nn.Parameter):
+            for n, p in self.root.named_parameters():
+                if attr_val is p:
+                    if n not in parameter_proxy_cache:
+                        proxy = self.create_proxy('get_attr', n, (), {})
+                        parameter_proxy_cache[n] = ProxyTensor(attr_val, proxy)
+                    return parameter_proxy_cache[n]
+            return attr_val
+        return attr_val
+
+    # We need to do this so that parameters entering the `make_fx` context have
+    # a reference to them (and also have requires_grad set on them correctly
+    # I'm not actually sure if this is the right thing to do ...
```

Comment:
can we figure this out lol

Reply:
I don't think we need to set requires_grad on it (at least, I don't remember why'd we need to and this code doesn't seem to do anything special about requires_grad...).

So I'll just remove the comment.

Reply:
I think the main complexity here is that fundamentally, the `.backward()` API allows you to "leak" things like tracing tensors.

For example, what should happen here?

```
model = resnet18()
def f(x):
    model(x).sum().backward()
    return x.grad

print(make_fx(f)(torch.randn(1,3,224,224, requires_grad=True)).code)

print([i.grad for i in model.parameters()])
```

Under the current implementation, this will leak ProxyTensors to outside the function. I think this is unavoidable, and just part of the contract of `make_fx`.

Reply:
Oh yeah, we talked about this on VC just now. It seems to me that when non-proxy `requires_grad=True` enter the scope of f, they should get detached and turned into leaf nodes, so the proxy doesn't escape. WDYT?